Sources/HTTP/Parser/CHTTPParser.swift

```diff
@@ -0,0 +1,231 @@
+import Transport
+import CHTTP
+import URI
+
+/// Internal CHTTP parser protocol
+internal protocol CHTTPParser: class {
+    associatedtype StreamType: ReadableStream
+    var stream: StreamType { get }
+    var parser: http_parser { get set }
+    var settings: http_parser_settings { get set }
+    var buffer: Bytes { get set }
+}
+
+extension CHTTPParser {
+    /// Parses a generic CHTTP message, filling the
+    /// ParseResults object attached to the C praser.
+    internal func parseMessage() throws -> ParseResults {
+        // create a new results object and set
+        // a reference to it on the parser
+        var results = ParseResults()
+        ParseResults.set(&results, on: &parser)
+        
+        // called when chunks of the url have been read
+        settings.on_url = { parser, chunk, length in
+            guard
+                let results = ParseResults.get(from: parser),
+                let bytes = chunk?.makeBytes(length: length)
+            else {
+                // signal an error
+                return 1
```

Comment:
I'd consider wrapping `1` and `0` in an enum for `failure` and `success` to help eliminate confusion for those not used to C error status.